src/platforms/esp32/components/avm_builtins/uart_driver.c

```diff
@@ -347,6 +349,19 @@ static void uart_driver_do_read(Context *ctx, GenMessage gen_message)
     }
 }
 
+static void uart_driver_do_cancel_read(Context *ctx, GenMessage gen_message)
+{
+    struct UARTData *uart_data = ctx->platform_data;
+
+    uart_data->reader_process_pid = term_invalid_term();
```

Comment:
`uart_interrupt_callback` can run on the other scheduler on dual core esp32s. So let's take advantage of this PR to protect uart_data accesses with a mutex or Atomic fields if it's sufficient.

Reply:
I think making the two fields `uart_data->reader_process_pid`, and `uart_data->reader_ref_ticks` atomic should mitigate the problem, since these are both touched during read, cancel_read, and the interrupt callback.